stdlib/public/core/Codable.swift

```diff
@@ -0,0 +1,3140 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// Codable
+//===----------------------------------------------------------------------===//
+
+/// Conformance to `Encodable` indicates that a type can encode itself to an external representation.
+public protocol Encodable {
+    /// Encodes `self` into the given encoder.
+    ///
+    /// If `self` fails to encode anything, `encoder` will encode an empty keyed container in its place.
+    ///
+    /// - parameter encoder: The encoder to write data to.
+    /// - throws: An error if any values are invalid for `encoder`'s format.
+    func encode(to encoder: Encoder) throws
```

Comment:
@rintaro `encode(to:)` and `init(from:)` are not generic because there are cases where you need to pass an existential `Encoder` or `Decoder` to an object. For instance, given:

```swift
class A : Codable { /* ... */ }
class B : A { /* ... */ }
```

`class B` would likely implement `encode(to:)` as

```swift
func encode(to encoder: Encoder) throws {
    var container = encoder./* some container type */
    // encode properties into container

    super.encode(to: container.superEncoder())
}
```

This is not possible to do if `encode(to:)` is generic, because `superEncoder()` cannot return a concrete type, and `encode<E : Encoder>(to encoder: E)` cannot take an existential type.